Lib/multiprocessing/popen_spawn_win32.py

```diff
@@ -44,6 +46,19 @@ class Popen(object):
     def __init__(self, process_obj):
         prep_data = spawn.get_preparation_data(process_obj._name)
 
+        # bpo-42178: Check for pickle error before creating subprocess
+        self.sentinel = None
+        set_spawning_popen(self)
+        try:
+            with io.BytesIO() as f:
+                reduction.dump(prep_data, f)
+                try:
+                    reduction.pickle.dump(process_obj, f)
+                except:
+                    reduction.dump(process_obj, f)
+        finally:
+            set_spawning_popen(None)
+
```

Comment:
Pickling `Semaphore` or `SemLock` objects entails duplicating an OS handle to the child process via `duplicate_for_child`. So we need the child process in order to do the reduction only once. What we can do to improve the situation is to spawn the child process with a suspended thread. This will require adding the `CREATE_SUSPENDED` flag and `ResumeThread` function to the `_winapi` module.  After creating the child process with a suspended thread, set up the self attributes (e.g. `self.sentinel`) and dump the reduction to a `BytesIO` instance. If the latter fails, kill the child process via `_winapi.TerminateProcess`. Otherwise resume the thread in the child, and write the reduction from the `BytesIO` instance to the pipe.